executor/distsql_test.go

```diff
@@ -477,3 +480,78 @@ func TestCoprocessorPagingSize(t *testing.T) {
 	rpcNum = getRPCNumFromExplain(rows)
 	require.Equal(t, rpcNum, uint64(1))
 }
+
+func TestAdaptiveClosestRead(t *testing.T) {
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+
+	readCounter := func(counter prometheus.Counter) float64 {
+		var metric dto.Metric
+		require.Nil(t, counter.Write(&metric))
+		return metric.Counter.GetValue()
+	}
+
+	checkMetrics := func(q string, hit, miss int) {
+		beforeHit := readCounter(metrics.DistSQLCoprClosestReadHitCounter.WithLabelValues("hit"))
+		beforeMiss := readCounter(metrics.DistSQLCoprClosestReadHitCounter.WithLabelValues("miss"))
+		tk.MustQuery(q)
+		afterHit := readCounter(metrics.DistSQLCoprClosestReadHitCounter.WithLabelValues("hit"))
+		afterMiss := readCounter(metrics.DistSQLCoprClosestReadHitCounter.WithLabelValues("miss"))
+		require.Equal(t, hit, int(afterHit-beforeHit), "exec query '%s' check hit failed", q)
+		require.Equal(t, miss, int(afterMiss-beforeMiss), "exec query '%s' check miss failed", q)
+	}
+
+	tk.MustExec("create table t(id int primary key, s varchar(8), p varchar(16));")
+	// TODO: unistore does not implement region split, this is just a noop.
+	tk.MustExec("split table t by (3);")
+	tk.MustExec("insert into t values (1, '00000001', '0000000000000001'), (2, '00000003', '0000000000000002'), (3, '00000011', '0000000000000003');")
+	tk.MustExec("analyze table t;")
+
+	tk.MustExec("set tidb_replica_read = 'closest-adaptive';")
+	tk.MustExec("set tidb_adaptive_closest_read_threshold = 25;")
+
+	// table reader
+	// estimate cost is 19
+	checkMetrics("select s from t where id >= 1 and id < 2;", 0, 1)
+	// estimate cost is 37
+	checkMetrics("select * from t where id >= 1 and id < 2;", 1, 0)
+	// estimate cost is 74
+	checkMetrics("select * from t where id >= 1 and id <= 2;", 1, 0)
+	// FIXME: there is only 1 region in unistore, can't test multi region read here.
+	// checkMetrics("select * from t where id >= 2 and id < 4;", counter.WithLabelValues("miss"), 2)
```

Comment:
Can we add this case in the dir `tests/realtikvtest/`

Reply:
I tried with realtikv, but found the `split table regions` is too slow, after 10s there was still only one region. Add cases to test partition table instead.